bench.cpp

```diff
@@ -1,99 +1,113 @@
-      "Data array size, ususally a column size in elements");
-  desc.add_options()("iterations", po::value<size_t>(&opts->iterations),
-                     "Number of iterations to run a bmark.");
-  desc.add_options()("device",
-                     po::value<RunOptions::DeviceType>(&opts->device_ty),
-                     "Device to run on.");
-  desc.add_options()("report_path", po::value<std::string>(&opts->report_path),
-                     "Full/Relative path to a report file.");
-  desc.add_options()(
-      "groups_count", po::value<size_t>(&groups_count),
-      "Number of unique keys for dwarfs with keys (groupby, hash build etc.).");
-  desc.add_options()("executors", po::value<size_t>(&executors),
-                     "Number of executors for GroupByLocal.");
-  po::positional_options_description pos_opts;
-  pos_opts.add("dwarf", 1);
-
-  po::variables_map vm;
-
-  try {
-    po::store(po::command_line_parser(argc, argv)
-                  .options(desc)
-                  .positional(pos_opts)
-                  .run(),
-              vm);
-    vm.notify();
-
-    if (dwarf_name == "list") {
-      std::cout << "Supported dwarfs:\n";
-      for (const auto &dw : *registry) {
-        std::cout << "\t" << dw.first << std::endl;
-      }
-      return 0;
-    } else {
-      dwarf = registry->find(dwarf_name);
-    }
-    if (vm.count("help")) {
-      std::cout << desc;
-      return 0;
-    } else if (!dwarf) {
-      std::cerr << "List supported dwarfs to run with '" << argv[0] << " list'"
-                << std::endl;
-      return 1;
-    }
-
-    if (opts->input_size.empty()) {
-      opts->input_size.push_back(1);
-    }
-
-    helpers::set_dpcpp_filter_env(*opts);
-
-    if (isGroupBy(dwarf_name)) {
-      std::unique_ptr<GroupByRunOptions> tmpPtr =
-          std::make_unique<GroupByRunOptions>(*opts, groups_count, executors);
-      opts.reset();
-      opts = std::move(tmpPtr);
-    }
-
-    dwarf->init(*opts);
-    dwarf->run(*opts);
-    dwarf->report(*opts);
-  } catch (std::exception &e) {
-    std::cerr << "Caught exception: " << e.what() << std::endl;
+std::vector<Measurement> DwarfBench::makeMeasurements(const RunConfig &conf) {
+  static Registry *reg = []() {
+    populate_registry();
+    return Registry::instance();
+  }();
+
+  RunOptions _opts = RunOptions{.device_ty = conf.device == DeviceType::CPU
+                                                 ? RunOptions::DeviceType::CPU
+                                                 : RunOptions::DeviceType::GPU,
+                                .input_size = {conf.inputSize},
+                                .iterations = conf.iterations,
+                                .report_path = ""};
+
+  GroupByRunOptions opts = GroupByRunOptions(_opts, 20, 1024); // TODO
+
+  std::string dwarfName = dwarfToString(conf.dwarf);
+  auto dwarf = reg->find(dwarfName);
+  if (dwarf == nullptr) {
+    throw DwarfBenchException("Internal error: Wrong dwarf name: `" +
```

Comment:
This should be an invalid parameter exception if we allow passing arbitrary dwarfs. However, since we don't, this should not be an exception in the first place. What should a user do with such an exception?

Reply:
Should we take into account that we may forget to add a new dwarf to `dwarfToString` function? Or should it be done differently?

Reply:
Not really, since you'll catch that on the assertion